codex/contracts/proofs.nim

```diff
@@ -22,6 +22,14 @@ method periodicity*(proofs: OnChainProofs): Future[Periodicity] {.async.} =
   let period = await proofs.storage.proofPeriod()
   return Periodicity(seconds: period)
 
+method isSlotCancelled*(proofs: OnChainProofs,
```

Comment:
> Proofs and Marketplace are two separate abstractions that get mixed here. Marketplace concerns itself with requests and slots, whereas Proofs concerns itself with proof requirements.
Definitely agreed. In `dagger-contracts` the abstraction is maintained and `isSlotCancelled`/`isCancelled` live in `storage.sol`.

>  I would suggest to update isProofRequired(), willProofBeRequired() and proofEnd() in Solidity to take the cancelled state into account. Then there's probably no need to update the Nim code around proofs much.

This was done in https://github.com/status-im/dagger-contracts/pull/19, although I'm not sure `proofEnd` was considered as it returns a timestamp. Should it be updated such that `proofEnd` returns `block.timestamp` if the request/slot was cancelled/finished?

Reply:
> I do see why you'd need to check whether a slot is cancelled in order to stop requiring proofs, but I would suggest to update isProofRequired(), willProofBeRequired() and proofEnd() in Solidity to take the cancelled state into account. Then there's probably no need to update the Nim code around proofs much.

The downside to this approach is that the proofs for cancelled contracts will continue to be checked but will not be required because `isProofRequired` will be false. Removing the slot from `proving.slots` will be more efficient, but we need to figure out whether or not the slot was cancelled.

Reply:
In https://github.com/status-im/dagger-contracts/pull/19, I'll update `Marketplace.proofEnd` to use the `slotAcceptsProofs(slotId)` modifier (introduced here: https://github.com/status-im/dagger-contracts/pull/19/files#diff-5973808e32384782b5978a63e85b893b67d634be8aa42001f0ef480f189a0688R348) and omit the `removeCancelledContracts` change in `proving.nim` as calling `proofEnd` should revert. Will add tests for that as well.